File: 138-Copy List with Random Pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x, next=None, random=None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: Node
        :rtype: Node
        """
# first try:

        if not head:
            return None
        hmap = dict()
        tmp = newHead = Node(head.val, None, None)
        dummyHead = head
        
        while dummyHead:
            hmap[dummyHead] = tmp
            dummyHead = dummyHead.next
            if dummyHead:
                tmp.next = Node(dummyHead.val, None, None)
                tmp = tmp.next
        
        tmp = newHead
        
        while tmp:
            if head.random:
                tmp.random = hmap[head.random]
            else:
                tmp.random = None
            tmp = tmp.next
            head = head.next
            
        return newHead
    
# recursive solution
        hmap = dict()
    
        def recurse(node):
            if not node:
                return None
            if node in hmap:
                return hmap[node]
            copyNode = Node(node.val, None, None)
            hmap[node] = copyNode
            copyNode.next = recurse(node.next)
            copyNode.random = recurse(node.random)
            return copyNode
        
        return recurse(head)

File: JavaScript/138-Copy List with Random Pointer.js
/**
 * // Definition for a Node.
 * function Node(val, next, random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */

/**
 * @param {Node} head
 * @return {Node}
 */
var copyRandomList = function(head) {
    let hmap = new Map();
    
    let recurse = (node) =>{
        if (!node) return null;
        if (hmap.has(node)) return hmap.get(node);
        let copyNode = new Node(node.val, null, null);
        hmap.set(node, copyNode);
        copyNode.next = recurse(node.next);
        copyNode.random = recurse(node.random);
        return copyNode;
    }
    
    return recurse(head);
};